File: app/src/main/java/com/dmitriykargashin/cardamontimecalculator/engine/calculator/CalculatorOfTime.kt
/*
 * Copyright (c) 2019. Dmitriy Kargashin
 */

package com.dmitriykargashin.cardamontimecalculator.engine.calculator

import android.util.Log
import com.dmitriykargashin.cardamontimecalculator.data.tokens.Token
import com.dmitriykargashin.cardamontimecalculator.data.tokens.TokenType
import com.dmitriykargashin.cardamontimecalculator.data.tokens.Tokens
import com.dmitriykargashin.cardamontimecalculator.utilites.TimeConverter
import net.objecthunter.exp4j.ExpressionBuilder


abstract class CalculatorOfTime {


    companion object {


        fun evaluate(tokensToEvaluate: Tokens): Tokens {

            val clonedTokensToEvaluate = tokensToEvaluate.clone()
            //

            return if (clonedTokensToEvaluate.isSimpleArithmeticExpression())
                evaluateSimpleArithmeticExpression(clonedTokensToEvaluate)
            else {
                val tokensWithParentheses = setParenthesesToExpression(clonedTokensToEvaluate)
                val tokensinMsecs = TimeConverter.convertExpressionToMsecs(tokensWithParentheses)

                //  return convertExpressionInMsecsToType(evaluatedToken[0], TokenType.HOUR)
                evaluateSimpleArithmeticExpression(tokensinMsecs)//TimeConverter.convertExpressionInMsecsToNearest(evaluatedToken[0])
            }


        }


        private fun evaluateSimpleArithmeticExpression(tokensToEvaluate: Tokens): Tokens {

            Log.i("evaluate before", tokensToEvaluate.toString())

//if we have trailing operator in expression we need to delete it
            if (tokensToEvaluate.lastIndex >= 0 && tokensToEvaluate.last().type.isOperator()) {
                tokensToEvaluate.removeLastToken()
            }

            if (tokensToEvaluate.lastIndex >= 1 && tokensToEvaluate.last().type == TokenType.PARENTHESESRIGHT
                && tokensToEvaluate.elementAt(tokensToEvaluate.lastIndex - 1).type.isOperator())
             {
                tokensToEvaluate.removeLastToken().removeLastToken()
            }


            val txt = tokensToEvaluate.toString()
            Log.i("evaluate", txt)

            val resultTokens = Tokens()
            if (txt == "") return resultTokens // expression is empty so return empty Tokens list

            // Create an Expression (A class from exp4j library)

            try {
                val expression = ExpressionBuilder(txt).build()
                // Calculate the result
                val result = expression.evaluate().toBigDecimal()

                // we'll return result as one NUMBER token

             /*   val fmt = NumberFormat.getInstance()
                fmt.isGroupingUsed = false
                fmt.maximumIntegerDigits = 999
                fmt.maximumFractionDigits = 999
                val resultAsString = fmt.format(result)*/
                val resultAsString=result.toString()

                Log.i("result", result.toString())
                resultTokens.add(
                    Token(
                        TokenType.NUMBER,
                        resultAsString
                        /*result.toString()*/
                    )
                )

                resultTokens.add(
                    Token(
                        TokenType.MSECOND

                    )
                )
                return resultTokens

                // txtInput.text = result.toString()
                //     lastDot = true // Result contains a dot
            } catch (ex: Exception) {
                //   val resultTokens = Tokens()
                resultTokens.add(
                    Token(
                        TokenType.ERROR,
                        "ERROR"
                    )
                )
                Log.i("TAG", "Catch ERROR")
                return resultTokens

            }

        }





        private fun setParenthesesToExpression(tokensToSetParntheses: Tokens): Tokens {
            val tokensWithParentheses = Tokens()
            var isParenthesesBegins = false
            for (token in tokensToSetParntheses) {
                when (token.type) {
                    TokenType.NUMBER -> {
                        if (!isParenthesesBegins) {
                            tokensWithParentheses.add(
                                Token(
                                    TokenType.PARENTHESESLEFT
                                )
                            )
                            isParenthesesBegins = true
                        }
                    }
                    TokenType.MULTIPLY, TokenType.DIVIDE, TokenType.MINUS, TokenType.PLUS -> {

                        tokensWithParentheses.add(Token(TokenType.PARENTHESESRIGHT))
                        isParenthesesBegins = false

                    }
                }
                tokensWithParentheses.add(
                    Token(
                        token.type,
                        token.strRepresentation
                    )
                )

            }
            tokensWithParentheses.add(Token(TokenType.PARENTHESESRIGHT))
            return tokensWithParentheses
        }

    }

    //todo -In result of negative expression will be terms with MINUS symbol EACH

}

File: flutter_app/lib/engine/calculator_of_time.dart
import 'big_decimal.dart';
import 'expression_evaluator.dart';
import 'time_converter.dart';
import 'token.dart';
import 'token_type.dart';
import 'tokens.dart';

/// The calculation engine. Port of Kotlin `CalculatorOfTime`
/// (engine/calculator/CalculatorOfTime.kt).
abstract final class CalculatorOfTime {
  /// Evaluates a token expression. The result is always one of:
  /// * an empty [Tokens] (empty input),
  /// * `[NUMBER <milliseconds as a double string>, MSECOND]` on success -
  ///   the MSECOND tag is appended even for pure-number expressions and
  ///   carries the msec total as its value (so it reads "MSeconds" unless
  ///   the total is exactly 1),
  /// * `[ERROR "ERROR"]` on any failure (malformed input, division by zero).
  ///
  /// Time expressions are rewritten before evaluation: each maximal
  /// `NUMBER [unit] ...` run is parenthesized, each unit becomes
  /// `* <milliseconds-per-unit>` and each number gets a leading `+`, so
  /// `5 Hour - 10 Minute` evaluates `(+5*3600000)-(+10*60000)`.
  static Tokens evaluate(Tokens tokensToEvaluate) {
    final clonedTokensToEvaluate = tokensToEvaluate.clone();
    if (clonedTokensToEvaluate.isSimpleArithmeticExpression()) {
      return _evaluateSimpleArithmeticExpression(clonedTokensToEvaluate);
    }
    final tokensWithParentheses =
        _setParenthesesToExpression(clonedTokensToEvaluate);
    final tokensInMsecs =
        TimeConverter.convertExpressionToMsecs(tokensWithParentheses);
    return _evaluateSimpleArithmeticExpression(tokensInMsecs);
  }

  static Tokens _evaluateSimpleArithmeticExpression(Tokens tokensToEvaluate) {
    // If we have a trailing operator in the expression we need to delete it.
    if (tokensToEvaluate.isNotEmpty && tokensToEvaluate.last.type.isOperator) {
      tokensToEvaluate.removeLastToken();
    }
    // Repair a trailing "<operator> )" pair produced by the parenthesizer.
    if (tokensToEvaluate.length >= 2 &&
        tokensToEvaluate.last.type == TokenType.parenthesesRight &&
        tokensToEvaluate[tokensToEvaluate.length - 2].type.isOperator) {
      tokensToEvaluate.removeLastToken().removeLastToken();
    }

    final txt = tokensToEvaluate.toString();
    final resultTokens = Tokens();
    if (txt == '') return resultTokens;

    try {
      final result = ExpressionEvaluator.evaluate(txt);
      // Kotlin's Double.toBigDecimal() throws on Infinity/NaN inside the
      // same try block; mirror that so non-finite results become ERROR.
      if (!result.isFinite) {
        throw const FormatException('Non-finite result');
      }
      // Dart's double.toString() is used here where Kotlin used
      // Double.toBigDecimal().toString(); the string differs in shape
      // ("18600000.0" vs "1.86E+7") but is internal: every consumer parses
      // it with BigDecimal.parse and re-emits plain strings.
      final resultAsString = result.toString();
      // RemoveADS value plumbing: both the NUMBER and the MSECOND tag carry
      // the msec total, so the tag renders "MSeconds" unless the total is
      // exactly 1 (display-only; downstream matches by type).
      final resultValue = BigDecimal.parse(resultAsString);
      resultTokens.add(Token(TokenType.number, resultValue, resultAsString));
      resultTokens.add(Token(TokenType.mSecond, resultValue));
      return resultTokens;
    } catch (_) {
      resultTokens.add(Token(TokenType.error, BigDecimal.one, 'ERROR'));
      return resultTokens;
    }
  }

  static Tokens _setParenthesesToExpression(Tokens tokensToSetParentheses) {
    final tokensWithParentheses = Tokens();
    var isParenthesesBegins = false;
    for (final token in tokensToSetParentheses) {
      switch (token.type) {
        case TokenType.number:
          if (!isParenthesesBegins) {
            tokensWithParentheses
                .add(Token(TokenType.parenthesesLeft, BigDecimal.one));
            isParenthesesBegins = true;
          }
        case TokenType.multiply:
        case TokenType.divide:
        case TokenType.minus:
        case TokenType.plus:
          tokensWithParentheses
              .add(Token(TokenType.parenthesesRight, BigDecimal.one));
          isParenthesesBegins = false;
        default:
          break;
      }
      // Verbatim copy (keeps token.value; the explicit strRepresentation is
      // never re-pluralized - the branch grew an extra 's' here, see Token).
      tokensWithParentheses
          .add(Token(token.type, token.value, token.strRepresentation));
    }
    tokensWithParentheses
        .add(Token(TokenType.parenthesesRight, BigDecimal.one));
    return tokensWithParentheses;
  }
}
